In [1]:
import numpy as np
import pandas as pd

In [2]:
# Задаем рабочую папку

import os
os.chdir("C:/Users/Aleksandr/Desktop/CSC/Data Analysis term 2/lab 2")

In [3]:
# Импорт данных
wine = pd.read_csv('winequality-white.csv', sep=';', header=0)
wine.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [4]:
#  Разделяем предикторы и отклик
# Отклик - группирующая переменная -  вектор y
y = wine['quality']
# Предикторы - таблица X
X = wine.drop('quality', axis=1)

In [5]:
#отклик не сбалансирован

In [6]:
#запишем все вина 3й категории в 4ю, а 9й  в 8ю (они похожи по качеству)
for i in range(1,4897):
    if (wine.at[i, 'quality'] == 3):
        wine.at[i, 'quality'] = 4
    if (wine.at[i, 'quality'] == 9):
        wine.at[i, 'quality'] = 8

In [7]:
#  Отклик сбалансирован? Теперь да
wine['quality'].value_counts(normalize=True)

6    0.448755
5    0.297468
7    0.179665
4    0.037362
8    0.036750
Name: quality, dtype: float64

In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=0.33)

In [9]:
from xgboost import XGBClassifier

from sklearn import metrics

In [12]:
from sklearn.model_selection import GridSearchCV

In [105]:
grid_param = {  
    'n_estimators': [100], #проверил на 50, 100, 500, 1000, 2000
    'max_depth': [8, 10, 15],
    'cv': [3], #3 хорошо, проверил на 2, 3, 5, 10
    'gamma': [0, 0.1],
    'learning_rate': [0.01, 0.1] #проверил на 0.01, 0.1, 0.2
    # 'eta' = [0.01, 0.1, 0.2] связана с learning rate
}

In [106]:
model_grid = GridSearchCV(
    error_score='raise',
    estimator=XGBClassifier(
            seed=42,
            eta = 0.2,
            n_estimators=1000,
            max_depth=4,
            learning_rate=0.1,
            #base_score=0.5, 
            #seed=1234, 
            objective='binary:logistic',
            cv=3,
            gamma=0, 
            #learning_rate=0.1, 
            max_delta_step=0,
            min_child_weight=1,
            missing=None, 
            nthread=-1,
            #reg_alpha=0, 
            #reg_lambda=1,
            #scale_pos_weight=1, 
            silent=False,
            n_jobs = -1       
    ),
    #iid=True, 
    param_grid=grid_param, 
    #refit=True, 
    scoring='accuracy', 
    verbose=3)

In [107]:
model_grid.fit(X_train, y_train)

C:\Users\Aleksandr\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Fitting 3 folds for each of 12 candidates, totalling 36 fits
[CV] cv=3, gamma=0, learning_rate=0.01, max_depth=8, n_estimators=100 
[CV]  cv=3, gamma=0, learning_rate=0.01, max_depth=8, n_estimators=100, score=0.605, total=   1.4s
[CV] cv=3, gamma=0, learning_rate=0.01, max_depth=8, n_estimators=100 


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    1.3s remaining:    0.0s


[CV]  cv=3, gamma=0, learning_rate=0.01, max_depth=8, n_estimators=100, score=0.599, total=   1.4s
[CV] cv=3, gamma=0, learning_rate=0.01, max_depth=8, n_estimators=100 


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    2.7s remaining:    0.0s


[CV]  cv=3, gamma=0, learning_rate=0.01, max_depth=8, n_estimators=100, score=0.567, total=   1.4s
[CV] cv=3, gamma=0, learning_rate=0.01, max_depth=10, n_estimators=100 
[CV]  cv=3, gamma=0, learning_rate=0.01, max_depth=10, n_estimators=100, score=0.628, total=   1.8s
[CV] cv=3, gamma=0, learning_rate=0.01, max_depth=10, n_estimators=100 
[CV]  cv=3, gamma=0, learning_rate=0.01, max_depth=10, n_estimators=100, score=0.619, total=   1.7s
[CV] cv=3, gamma=0, learning_rate=0.01, max_depth=10, n_estimators=100 
[CV]  cv=3, gamma=0, learning_rate=0.01, max_depth=10, n_estimators=100, score=0.569, total=   1.7s
[CV] cv=3, gamma=0, learning_rate=0.01, max_depth=15, n_estimators=100 
[CV]  cv=3, gamma=0, learning_rate=0.01, max_depth=15, n_estimators=100, score=0.616, total=   2.3s
[CV] cv=3, gamma=0, learning_rate=0.01, max_depth=15, n_estimators=100 
[CV]  cv=3, gamma=0, learning_rate=0.01, max_depth=15, n_estimators=100, score=0.609, total=   2.4s
[CV] cv=3, gamma=0, learning_rate=0.01, m

[Parallel(n_jobs=1)]: Done  36 out of  36 | elapsed:  1.0min finished


GridSearchCV(cv='warn', error_score='raise',
             estimator=XGBClassifier(base_score=0.5, booster='gbtree',
                                     colsample_bylevel=1, colsample_bynode=1,
                                     colsample_bytree=1, cv=3, eta=0.2, gamma=0,
                                     learning_rate=0.1, max_delta_step=0,
                                     max_depth=4, min_child_weight=1,
                                     missing=None, n_estimators=1000, n_jobs=-1,
                                     nthread=-1, objective='binary:logistic',
                                     random_state=0, reg_alpha=0, reg_lambda=1,
                                     scale_pos_weight=1, seed=42, silent=False,
                                     subsample=1, verbosity=1),
             iid='warn', n_jobs=None,
             param_grid={'cv': [3], 'gamma': [0, 0.1],
                         'learning_rate': [0.01, 0.1], 'max_depth': [8, 10, 15],
                        

In [108]:
# summarize results
print("Best: %f using %s" % (model_grid.best_score_, model_grid.best_params_))
means = model_grid.cv_results_['mean_test_score']
params = model_grid.cv_results_['params']
for mean, param in zip(means, params):
    print("%f with: %r" % (mean, param))

Best: 0.629991 using {'cv': 3, 'gamma': 0, 'learning_rate': 0.1, 'max_depth': 10, 'n_estimators': 100}
0.590369 with: {'cv': 3, 'gamma': 0, 'learning_rate': 0.01, 'max_depth': 8, 'n_estimators': 100}
0.605608 with: {'cv': 3, 'gamma': 0, 'learning_rate': 0.01, 'max_depth': 10, 'n_estimators': 100}
0.598598 with: {'cv': 3, 'gamma': 0, 'learning_rate': 0.01, 'max_depth': 15, 'n_estimators': 100}
0.626029 with: {'cv': 3, 'gamma': 0, 'learning_rate': 0.1, 'max_depth': 8, 'n_estimators': 100}
0.629991 with: {'cv': 3, 'gamma': 0, 'learning_rate': 0.1, 'max_depth': 10, 'n_estimators': 100}
0.626333 with: {'cv': 3, 'gamma': 0, 'learning_rate': 0.1, 'max_depth': 15, 'n_estimators': 100}
0.591588 with: {'cv': 3, 'gamma': 0.1, 'learning_rate': 0.01, 'max_depth': 8, 'n_estimators': 100}
0.606522 with: {'cv': 3, 'gamma': 0.1, 'learning_rate': 0.01, 'max_depth': 10, 'n_estimators': 100}
0.598598 with: {'cv': 3, 'gamma': 0.1, 'learning_rate': 0.01, 'max_depth': 15, 'n_estimators': 100}
0.618409 with: 

In [115]:
#Создаю модель с лучшими параметрами:
best_model = XGBClassifier(
            seed=42,
            #eta = 0.2,
            n_estimators=100,
            max_depth=10,
            learning_rate=0.1,
            objective='binary:logistic',
            cv=3,
            gamma=0, 
            missing=None, 
            nthread=-1,
            silent=False,
            n_jobs = -1       
    )

best_model.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, cv=3, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=10,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=-1,
              nthread=-1, objective='multi:softprob', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=42,
              silent=False, subsample=1, verbosity=1)

In [116]:
# Оценка качества модели
y_pred_train = best_model.predict(X_train)

y_pred_test = best_model.predict(X_test)

print("Accuracy on train set is ", metrics.accuracy_score(y_train, y_pred_train)*100)
print("Accuracy on test set is ", metrics.accuracy_score(y_test, y_pred_test)*100)

Accuracy on train set is  99.96952148735141
Accuracy on test set is  66.17192331478046


И я не очень понимаю, почему во время gridsearch score не превышал 0.65, зато если же взять модель с параметрами из gridsearch, то она на том же самом X_train, y_train выдает почти 100% acc. Куда девается 30% accuracy в гридсерче? 

И получается, что по гридсерчу - никакого переобучения нет, зато если запустить вручную проверку, то переобучение на лицо :(

И можно смело упростить модель

In [139]:
#Создаю модель с упрощенными параметрами:
simple_model = XGBClassifier(
            seed=42,
            learning_rate = 0.1,
            n_estimators=100,
            max_depth=5,
            cv=3,
            gamma=0, 
            nthread=-1,
            silent=False,       
    )

simple_model.fit(X_train, y_train)

# Оценка качества модели
y_pred_train_simple = simple_model.predict(X_train)

y_pred_test_simple = simple_model.predict(X_test)

print("Accuracy on train set is ", metrics.accuracy_score(y_train, y_pred_train_simple)*100)
print("Accuracy on test set is ", metrics.accuracy_score(y_test, y_pred_test_simple)*100)

Accuracy on train set is  82.71868332825359
Accuracy on test set is  60.17316017316018


In [161]:
max_score = 0

In [164]:
n_estimators = 1

for n_estimators in range(1,500,5):
    for max_depth in range(3,15,1):
        simple_model = XGBClassifier(
                seed = 42,
                learning_rate = 0.1,
                n_estimators = n_estimators,
                max_depth = max_depth,
                cv = 3,
                gamma = 0, 
                nthread = -1,
                silent = False,       
        )

        simple_model.fit(X_train, y_train)

        # Оценка качества модели
        y_pred_train_simple = simple_model.predict(X_train)
        y_pred_test_simple = simple_model.predict(X_test)

        #print("Accuracy on train set is ", metrics.accuracy_score(y_train, y_pred_train_simple)*100)
        #print("Accuracy on test set is ", metrics.accuracy_score(y_test, y_pred_test_simple)*100)
        
        score_train = metrics.accuracy_score(y_train, y_pred_train_simple)*100
        score_test = metrics.accuracy_score(y_test, y_pred_test_simple)*100
        print(n_estimators, max_depth, score_train, score_test, "\n")
        if (score_test > max_score):
            max_score = score_test


1 3 56.476683937823836 53.370439084724794 

1 4 58.85400792441329 54.483611626468765 

1 5 62.26760134105456 55.287569573283854 

1 6 65.34593111856142 53.865182436611 

1 7 69.85675099055166 54.112554112554115 

1 8 74.15422127400183 55.22572665429808 

1 9 76.40963120999696 54.916512059369204 

1 10 80.15848826577263 56.4625850340136 

1 11 83.05394696738799 56.95732838589982 

1 12 84.91313623895154 57.45207173778603 

1 13 85.91892715635477 56.77179962894249 

1 14 86.98567509905517 57.32838589981447 

6 3 56.87290460225542 53.555967841682126 

6 4 59.76836330387078 53.98886827458256 

6 5 64.18774763791527 55.04019789734076 

6 6 68.85096007314843 56.153370439084725 

6 7 74.61139896373057 57.14285714285714 

6 8 79.1526973483694 57.7612863327149 

6 9 85.1569643401402 59.0599876314162 

6 10 88.78390734532154 60.544217687074834 

6 11 91.58793050899116 61.16264687693259 

6 12 93.1118561414203 61.6573902288188 

6 13 94.39195367266078 61.224489795918366 

6 14 95.18439500152392 6

76 14 100.0 65.18243661100804 

81 3 66.4431575739104 57.14285714285714 

81 4 72.99603779335568 58.00865800865801 

81 5 80.46327339225846 59.43104514533086 

81 6 88.44864370618714 61.84291898577613 

81 7 94.6967387991466 63.821892393320965 

81 8 98.14081072843646 65.0587507730365 

81 9 99.39042974702834 66.54298082869512 

81 10 99.90856446205424 65.98639455782312 

81 11 100.0 65.73902288188003 

81 12 100.0 65.0587507730365 

81 13 100.0 65.3061224489796 

81 14 100.0 65.12059369202227 

86 3 66.68698567509905 56.95732838589982 

86 4 73.3313014324901 57.7612863327149 

86 5 81.01188661993295 59.863945578231295 

86 6 89.45443462359037 62.02844774273346 

86 7 95.27583053946968 64.19294990723562 

86 8 98.41511734227369 65.0587507730365 

86 9 99.48186528497409 66.11008039579468 

86 10 99.93904297470283 66.04823747680891 

86 11 100.0 65.61533704390847 

86 12 100.0 64.93506493506493 

86 13 100.0 65.18243661100804 

86 14 100.0 65.36796536796537 

91 3 67.14416336482779 57.14

166 7 99.48186528497409 65.42980828695114 

166 8 99.8171289241085 65.92455163883734 

166 9 100.0 65.36796536796537 

166 10 100.0 66.17192331478046 

166 11 100.0 65.73902288188003 

166 12 100.0 65.42980828695114 

166 13 100.0 64.87322201607915 

166 14 100.0 65.49165120593692 

171 3 71.53306918622371 57.88497217068645 

171 4 80.95092959463578 59.92578849721707 

171 5 90.27735446510209 62.21397649969078 

171 6 96.95214873514173 64.81137909709338 

171 7 99.51234379762268 65.18243661100804 

171 8 99.84760743675709 65.92455163883734 

171 9 100.0 65.5534941249227 

171 10 100.0 66.17192331478046 

171 11 100.0 65.61533704390847 

171 12 100.0 65.49165120593692 

171 13 100.0 65.24427952999382 

171 14 100.0 65.61533704390847 

176 3 71.74641877476378 57.7612863327149 

176 4 81.65193538555319 60.235003092145945 

176 5 90.6430966168851 62.27581941867656 

176 6 97.37884791222189 65.42980828695114 

176 7 99.51234379762268 65.49165120593692 

176 8 99.87808594940567 65.5534941249

KeyboardInterrupt: 

In [165]:
print(max_score)

66.54298082869512


Перебором удалось найти параметры с максимальным скором на тесовой выборке = 66.54